### ISSUE TITLE
Nx 13.0.1 release breaks nx-firebase use of createProjectGraph()

### ISSUE THREAD

**Issue by agalper** [simondotm/nx-firebase#44]
In recent release of Nx 13.0.1, createProjectGraph() is no longer a public API.

As a result the build executor fails for Firebase functions:

> nx run functions:build 
**project_graph_1.createProjectGraph is not a function**

Looks like createProjectGraphAsync() is still available. There may be more changes needed as well.

Any advice on working around this with Nx 13.0.1 is much appreciated.

**Comment by RandomDude123** [comment]
was able to make it run locally after updating to `13.1.2`.

- replace `createProjectGraph` with `createProjectGraphAsync`
- remove `nonBuildableDeps` logic in `nx-firebase/src/executors/build/build.ts` which fails during firebase:build

I don't manage to run the included e2e-tests locally and don't understand the logic behind the `nonBuildableDeps` checks and can't fix them. A PR seems therefore not really useful...

**Comment by edriang** [comment]
I'm having issues to, with version `13.1.4`

**Comment by simondotm** [comment]
Hi folks, I'm very sorry for the radio silence recently I had some personal issues to deal with.
I'll be able to look into this over the next week.

**Comment by tomschreck** [comment]
I tried @RandomDude123 recommendation in /node_modules/@simondotm/nx-firebase/src/executors/build/build.js.  Running NX's build command on project seems to get stuck at 'Copying asset files'.

Looking forward to a fix because getting Firebase functions to fit within NX monorepo would be awesome.

**Comment by BraunreutherA** [comment]
Hey there, 

I've created and PR: #45 - following @RandomDude123 work it was only missing to tell the `createProjectGraphAsync` to use the old version `4.0`. This is only a fix for now and the project should get refactored to NX V13. 

@simondotm I can put some effort into that an help if you already have a plan. Still I hope you can use this PR so we can go on using the plugin properly :)

**Comment by MrPumpking** [comment]
I've created a patch based on @BraunreutherA's PR:

`@simondotm+nx-firebase+0.3.3.patch`

```patch
diff --git a/node_modules/@simondotm/nx-firebase/src/executors/build/build.js b/node_modules/@simondotm/nx-firebase/src/executors/build/build.js
index ba14795..16aa79f 100644
--- a/node_modules/@simondotm/nx-firebase/src/executors/build/build.js
+++ b/node_modules/@simondotm/nx-firebase/src/executors/build/build.js
@@ -42,7 +42,7 @@ function runExecutor(options, context) {
         debugLog("Running Executor for Firebase Build for project '" + context.projectName + "'");
         debugLog('options=', options);
         // get the project graph; returns an object containing all nodes in the workspace, files, and dependencies
-        const projGraph = project_graph_1.createProjectGraph();
+        const projGraph = yield project_graph_1.createProjectGraphAsync('4.0');
         // nx firebase functions are essentially @nrwl/node:package libraries, but are added to the project
         // as applications as they are fundamentally the deployable "application" output of a build pipeline.
         // Due to this, we can import standard node libraries as dependencies from within the workspace
diff --git a/node_modules/@simondotm/nx-firebase/src/executors/build/node/package/package.js b/node_modules/@simondotm/nx-firebase/src/executors/build/node/package/package.js
index a3218e7..835ecb5 100644
--- a/node_modules/@simondotm/nx-firebase/src/executors/build/node/package/package.js
+++ b/node_modules/@simondotm/nx-firebase/src/executors/build/node/package/package.js
@@ -11,7 +11,7 @@ const normalize_options_1 = require("./utils/normalize-options");
 const cli_1 = require("./utils/cli");
 function packageExecutor(options, context) {
     return tslib_1.__awaiter(this, void 0, void 0, function* () {
-        const projGraph = project_graph_1.createProjectGraph();
+        const projGraph = yield project_graph_1.createProjectGraphAsync('4.0');
         const libRoot = context.workspace.projects[context.projectName].root;
         const normalizedOptions = normalize_options_1.default(options, context, libRoot);
         const { target, dependencies } = buildable_libs_utils_1.calculateProjectDependencies(projGraph, context.root, context.projectName, context.targetName, context.configurationName);

```

You can use it with with [patch-package](https://www.npmjs.com/package/patch-package) until the issue is resolved

**Comment by shannonjohnstone** [comment]
Just wanted to check in and see if we are any closer to getting [#45](https://github.com/simondotm/nx-firebase/pull/45) or another solution merged

**Comment by tomschreck** [comment]
checking in as well to see about getting [#45](https://github.com/simondotm/nx-firebase/pull/45) merged.

**Comment by anandsathe67** [comment]
Using the latest (as of this moment) version of nx (13.4.0), the following patch works. Had to tweak a few more things in addition to @MrPumpking 's changes

diff --git a/node_modules/@simondotm/nx-firebase/src/executors/build/build.js b/node_modules/@simondotm/nx-firebase/src/executors/build/build.js
index ba14795..5a62360 100644
[@simondotm+nx-firebase+0.3.3.patch.txt](https://github.com/simondotm/nx-firebase/files/7775799/%40simondotm%2Bnx-firebase%2B0.3.3.patch.txt)

--- a/node_modules/@simondotm/nx-firebase/src/executors/build/build.js
+++ b/node_modules/@simondotm/nx-firebase/src/executors/build/build.js
@@ -42,7 +42,8 @@ function runExecutor(options, context) {
         debugLog("Running Executor for Firebase Build for project '" + context.projectName + "'");
         debugLog('options=', options);
         // get the project graph; returns an object containing all nodes in the workspace, files, and dependencies
-        const projGraph = project_graph_1.createProjectGraph();
+        //const projGraph = project_graph_1.createProjectGraph();
+	const projGraph = yield project_graph_1.createProjectGraphAsync('4.0');
         // nx firebase functions are essentially @nrwl/node:package libraries, but are added to the project
         // as applications as they are fundamentally the deployable "application" output of a build pipeline.
         // Due to this, we can import standard node libraries as dependencies from within the workspace
@@ -146,7 +147,7 @@ function runExecutor(options, context) {
         // to be local package references to the copies we made
         const functionsPackageFile = `${options.outputPath}/package.json`;
         debugLog("- functions PackageFile=" + functionsPackageFile);
-        const functionsPackageJson = workspace_1.readJsonFile(functionsPackageFile);
+        const functionsPackageJson = devkit_1.readJsonFile(functionsPackageFile);
         const functionsPackageDeps = functionsPackageJson.dependencies;
         if (functionsPackageDeps) {
             debugLog("- Updating local dependencies for Firebase functions package.json");
@@ -164,7 +165,7 @@ function runExecutor(options, context) {
                 }
             }
         }
-        fileutils_1.writeJsonFile(functionsPackageFile, functionsPackageJson);
+        devkit_1.writeJsonFile(functionsPackageFile, functionsPackageJson);
         devkit_1.logger.log("- Updated firebase functions package.json");
         debugLog("functions package deps = ", JSON.stringify(functionsPackageDeps, null, 3));
         // Final dep check before we compile for:
diff --git a/node_modules/@simondotm/nx-firebase/src/executors/build/node/package/package.js b/node_modules/@simondotm/nx-firebase/src/executors/build/node/package/package.js
index a3218e7..b67cf93 100644
--- a/node_modules/@simondotm/nx-firebase/src/executors/build/node/package/package.js
+++ b/node_modules/@simondotm/nx-firebase/src/executors/build/node/package/package.js
@@ -11,7 +11,8 @@ const normalize_options_1 = require("./utils/normalize-options");
 const cli_1 = require("./utils/cli");
 function packageExecutor(options, context) {
     return tslib_1.__awaiter(this, void 0, void 0, function* () {
-        const projGraph = project_graph_1.createProjectGraph();
+        //const projGraph = project_graph_1.createProjectGraph();
+        const projGraph = yield project_graph_1.createProjectGraphAsync('4.0');
         const libRoot = context.workspace.projects[context.projectName].root;
         const normalizedOptions = normalize_options_1.default(options, context, libRoot);
         const { target, dependencies } = buildable_libs_utils_1.calculateProjectDependencies(projGraph, context.root, context.projectName, context.targetName, context.configurationName);
diff --git a/node_modules/@simondotm/nx-firebase/src/executors/build/node/package/utils/cli.js b/node_modules/@simondotm/nx-firebase/src/executors/build/node/package/utils/cli.js
index db9311d..d25ba80 100644
--- a/node_modules/@simondotm/nx-firebase/src/executors/build/node/package/utils/cli.js
+++ b/node_modules/@simondotm/nx-firebase/src/executors/build/node/package/utils/cli.js
@@ -1,9 +1,10 @@
 "use strict";
 Object.defineProperty(exports, "__esModule", { value: true });
 const fileutils_1 = require("@nrwl/workspace/src/utilities/fileutils");
+const devkit_1 = require("@nrwl/devkit");
 const fs_extra_1 = require("fs-extra");
 function addCliWrapper(options, context) {
-    const packageJson = fileutils_1.readJsonFile(`${options.outputPath}/package.json`);
+    const packageJson = devkit_1.readJsonFile(`${options.outputPath}/package.json`);
     const binFile = `${options.outputPath}/index.bin.js`;
     fileutils_1.writeToFile(binFile, `#!/usr/bin/env node
 'use strict';
@@ -13,7 +14,7 @@ require('${packageJson.main}');
     packageJson.bin = {
         [context.projectName]: './index.bin.js',
     };
-    fileutils_1.writeJsonFile(`${options.outputPath}/package.json`, packageJson);
+    devkit_1.writeJsonFile(`${options.outputPath}/package.json`, packageJson);
 }
 exports.default = addCliWrapper;
 //# sourceMappingURL=cli.js.map
diff --git a/node_modules/@simondotm/nx-firebase/src/executors/build/node/package/utils/update-package-json.js b/node_modules/@simondotm/nx-firebase/src/executors/build/node/package/utils/update-package-json.js
index 533b391..f043b59 100644
--- a/node_modules/@simondotm/nx-firebase/src/executors/build/node/package/utils/update-package-json.js
+++ b/node_modules/@simondotm/nx-firebase/src/executors/build/node/package/utils/update-package-json.js
@@ -1,6 +1,6 @@
 "use strict";
 Object.defineProperty(exports, "__esModule", { value: true });
-const fileutils_1 = require("@nrwl/workspace/src/utilities/fileutils");
+const fileutils_1 = require("@nrwl/devkit");
 const path_1 = require("path");
 function updatePackageJson(options, context) {
     const mainFile = path_1.basename(options.main).replace(/\.[tj]s$/, '');

**Comment by enchorb** [comment]
Can confirm the original patch by @MrPumpking still gave an error (fileutils_1.readJsonFile is not a function) however @anandsathe67 's additions seem to have resolved that - functions deployed successfully!

**Comment by johngrimsey** [comment]
This saved my ass thanks @MrPumpking and @anandsathe67 !

**Comment by jesusdpp96** [comment]
@anandsathe67's patch works fine. Now, the problem is when you installs new package ,changes are lossing (at less with 'yarn add' happened this)

**Comment by johngrimsey** [comment]
@jesusdpp96 Use https://www.npmjs.com/package/patch-package

**Comment by amkoehler** [comment]
Thank you, @MrPumpking @anandsathe67 and @BraunreutherA for your work on this! With nx 13.4 we're using the patch with `patch-package` and a postinstall script. It's working well but I suggest anyone using the patch to keep an eye on #45 to be merged so that the patch is no longer needed.

**Comment by ProductOfAmerica** [comment]
> Hi folks, I'm very sorry for the radio silence recently I had some personal issues to deal with. I'll be able to look into this over the next week.

I donate if you fix <3

**Comment by seanaguinaga** [comment]
How do apply a patch from a text file?

**Comment by wstidolph** [comment]
@anandsathe67 could you repost your patch in code/text rather than rendered markdown so I could just save it as a diff patch file? (Thanks!)

**Comment by victorouse** [comment]
> @anandsathe67 could you repost your patch in code/text rather than rendered markdown so I could just save it as a diff patch file? (Thanks!)


`@simondotm+nx-firebase+0.3.3.patch`

```
diff --git a/node_modules/@simondotm/nx-firebase/src/executors/build/build.js b/node_modules/@simondotm/nx-firebase/src/executors/build/build.js
index ba14795..8909a9b 100644
--- a/node_modules/@simondotm/nx-firebase/src/executors/build/build.js
+++ b/node_modules/@simondotm/nx-firebase/src/executors/build/build.js
@@ -15,7 +15,6 @@ const normalize_options_1 = require("./node/package/utils/normalize-options");
 const cli_1 = require("./node/package/utils/cli");
 const workspace_1 = require("@nrwl/workspace");
 const fs_extra_1 = require("fs-extra");
-const fileutils_1 = require("@nrwl/workspace/src/utilities/fileutils");
 const ENABLE_DEBUG = false;
 function debugLog(...args) {
     if (ENABLE_DEBUG) {
@@ -42,7 +41,7 @@ function runExecutor(options, context) {
         debugLog("Running Executor for Firebase Build for project '" + context.projectName + "'");
         debugLog('options=', options);
         // get the project graph; returns an object containing all nodes in the workspace, files, and dependencies
-        const projGraph = project_graph_1.createProjectGraph();
+        const projGraph = yield project_graph_1.createProjectGraphAsync('4.0');
         // nx firebase functions are essentially @nrwl/node:package libraries, but are added to the project
         // as applications as they are fundamentally the deployable "application" output of a build pipeline.
         // Due to this, we can import standard node libraries as dependencies from within the workspace
@@ -144,9 +143,8 @@ function runExecutor(options, context) {
         const incompatibleNestedDeps = [];
         // rewrite references to library packages in the functions package.json
         // to be local package references to the copies we made
-        const functionsPackageFile = `${options.outputPath}/package.json`;
+        const functionsPackageJson = devkit_1.readJsonFile(functionsPackageFile);
         debugLog("- functions PackageFile=" + functionsPackageFile);
-        const functionsPackageJson = workspace_1.readJsonFile(functionsPackageFile);
         const functionsPackageDeps = functionsPackageJson.dependencies;
         if (functionsPackageDeps) {
             debugLog("- Updating local dependencies for Firebase functions package.json");
@@ -164,7 +162,7 @@ function runExecutor(options, context) {
                 }
             }
         }
-        fileutils_1.writeJsonFile(functionsPackageFile, functionsPackageJson);
+        devkit_1.writeJsonFile(functionsPackageFile, functionsPackageJson);
         devkit_1.logger.log("- Updated firebase functions package.json");
         debugLog("functions package deps = ", JSON.stringify(functionsPackageDeps, null, 3));
         // Final dep check before we compile for:
diff --git a/node_modules/@simondotm/nx-firebase/src/executors/build/node/package/package.js b/node_modules/@simondotm/nx-firebase/src/executors/build/node/package/package.js
index a3218e7..43a4dc7 100644
--- a/node_modules/@simondotm/nx-firebase/src/executors/build/node/package/package.js
+++ b/node_modules/@simondotm/nx-firebase/src/executors/build/node/package/package.js
@@ -11,7 +11,7 @@ const normalize_options_1 = require("./utils/normalize-options");
 const cli_1 = require("./utils/cli");
 function packageExecutor(options, context) {
     return tslib_1.__awaiter(this, void 0, void 0, function* () {
-        const projGraph = project_graph_1.createProjectGraph();
+        const projGraph = yield project_graph_1.createProjectGraphAsync('4.0');
         const libRoot = context.workspace.projects[context.projectName].root;
         const normalizedOptions = normalize_options_1.default(options, context, libRoot);
         const { target, dependencies } = buildable_libs_utils_1.calculateProjectDependencies(projGraph, context.root, context.projectName, context.targetName, context.configurationName);
diff --git a/node_modules/@simondotm/nx-firebase/src/executors/build/node/package/utils/cli.js b/node_modules/@simondotm/nx-firebase/src/executors/build/node/package/utils/cli.js
index db9311d..d116377 100644
--- a/node_modules/@simondotm/nx-firebase/src/executors/build/node/package/utils/cli.js
+++ b/node_modules/@simondotm/nx-firebase/src/executors/build/node/package/utils/cli.js
@@ -1,11 +1,11 @@
 "use strict";
 Object.defineProperty(exports, "__esModule", { value: true });
-const fileutils_1 = require("@nrwl/workspace/src/utilities/fileutils");
+const devkit_1 = require("@nrwl/devkit");
 const fs_extra_1 = require("fs-extra");
 function addCliWrapper(options, context) {
-    const packageJson = fileutils_1.readJsonFile(`${options.outputPath}/package.json`);
+    const packageJson = devkit_1.readJsonFile(`${options.outputPath}/package.json`);
     const binFile = `${options.outputPath}/index.bin.js`;
-    fileutils_1.writeToFile(binFile, `#!/usr/bin/env node
+    devkit_1.writeToFile(binFile, `#!/usr/bin/env node
 'use strict';
 require('${packageJson.main}');
 `);
@@ -13,7 +13,7 @@ require('${packageJson.main}');
     packageJson.bin = {
         [context.projectName]: './index.bin.js',
     };
-    fileutils_1.writeJsonFile(`${options.outputPath}/package.json`, packageJson);
+    devkit_1.writeJsonFile(`${options.outputPath}/package.json`, packageJson);
 }
 exports.default = addCliWrapper;
 //# sourceMappingURL=cli.js.map
diff --git a/node_modules/@simondotm/nx-firebase/src/executors/build/node/package/utils/update-package-json.js b/node_modules/@simondotm/nx-firebase/src/executors/build/node/package/utils/update-package-json.js
index 533b391..dd88238 100644
--- a/node_modules/@simondotm/nx-firebase/src/executors/build/node/package/utils/update-package-json.js
+++ b/node_modules/@simondotm/nx-firebase/src/executors/build/node/package/utils/update-package-json.js
@@ -1,19 +1,19 @@
 "use strict";
 Object.defineProperty(exports, "__esModule", { value: true });
-const fileutils_1 = require("@nrwl/workspace/src/utilities/fileutils");
+const devkit_1 = require("@nrwl/devkit");
 const path_1 = require("path");
 function updatePackageJson(options, context) {
     const mainFile = path_1.basename(options.main).replace(/\.[tj]s$/, '');
     const typingsFile = `${mainFile}.d.ts`;
     const mainJsFile = `${mainFile}.js`;
-    const packageJson = fileutils_1.readJsonFile(path_1.join(context.root, options.packageJson));
+    const packageJson = devkit_1.readJsonFile(path_1.join(context.root, options.packageJson));
     if (!packageJson.main) {
         packageJson.main = `${options.relativeMainFileOutput}${mainJsFile}`;
     }
     if (!packageJson.typings) {
         packageJson.typings = `${options.relativeMainFileOutput}${typingsFile}`;
     }
-    fileutils_1.writeJsonFile(`${options.outputPath}/package.json`, packageJson);
+    devkit_1.writeJsonFile(`${options.outputPath}/package.json`, packageJson);
 }
 exports.default = updatePackageJson;
 //# sourceMappingURL=update-package-json.js.map
```

**Comment by shannonjohnstone** [comment]
Just wanted to ask that since this issue has been open for so long and we have no idea of when it possibly will be fixed, has anyone considered forking this repo implementing that required fixes and publishing a new package rather than all the work arounds?

Honest question as if someone has consider it doing so I was wanting to know what for "cons" have lead to that not happening?

**Comment by romshiri** [comment]
Did anyone fix it? facing the same issue.

**Comment by jimlloyd** [comment]
Ideally @simondotm will fix and publish a new release, but I think several of us are unblocked by using the patches provided above. Using `patch-package` might seem daunting but it is actually quite easy. I'm using `nx@13.8.3` with Node 16 and things seem to be working fine.

**Comment by romshiri** [comment]
@jimlloyd Thanks. Used patch-pacakage and it worked fine. 
Now I wish I could find a way to make the functions to "hot reload" every time I make a change..

**Comment by jimlloyd** [comment]
@romshiri Yes that's a problem. It's aggravating because the hot reload does work some (most?) of the time, but I have spent some time debugging a problem longer than necessary because I had not noticed that the hot reload wasn't working. Looking at the project README I see this note:

[](https://github.com/simondotm/nx-firebase/issues/44#technical-notes)IMPORTANT: Note that whilst nx serve will be useful when changing existing functions code, due to tsc --watch being enabled, it will NOT correctly detect changes if additional library imports are added to the source code or changes are made to any imported libraries during a watched session. To remedy this, relaunch the emulators by running nx serve again.

Perhaps that spells out exactly when hot reload doesn't work.

**Comment by romshiri** [comment]
@jimlloyd For me it doesn't work at all. It's a bit a bummer to serve it again on every small change. 
I'm thinking as a (not so ideal) workaround to create a simple node.js project just for the "rapid development phase" and then copy everything to the function once complete.

**Comment by jvalentik** [comment]
Hello, I noticed  another issue at least on latest NX that 
[copyAssetsFiles is undefined](https://github.com/simondotm/nx-firebase/blob/main/packages/nx-firebase/src/executors/build/build.ts#L92)
Here is really quick dirt hack added to @ben-kn-app addition to the original patch...
[@simondotm+nx-firebase+0.3.3.txt](https://github.com/simondotm/nx-firebase/files/8629459/%40simondotm%2Bnx-firebase%2B0.3.3.txt)

If anyone else stumbles upon this, just rename the file to .patch, drop it to patches folder and apply as per patch-package...
Thanks everyone here 👍

**Comment by adamstret** [comment]
So how about sharing a [petition for an official Firebase/Nx plugin](https://chng.it/wzG7GBBGgh)

**Comment by dpwanjala** [comment]
Thank you @jvalentik! Faced the [copyAssetsFiles is undefined](https://github.com/simondotm/nx-firebase/blob/main/packages/nx-firebase/src/executors/build/build.ts#L92) issue while using nx version 14.7.5 and your patch helped fix it.

**Comment by simondotm** [comment]
Finally released v0.3.4 of the plugin to npm with this fix. Apologies for how long this took to get done.

**Comment by simondotm** [comment]
Thanks to everyone for their contributions btw. 👍
